### PR TITLE
Migrations order fix

### PR DESCRIPTION
**Pull request by devgen-io**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | -
| Fixed issues  | #12974

### CODE REVIEW

**Comment by TerraSkye** [comment]
Wouldnt a uksort be better? Instead of creating a temp array?

File: framework/console/controllers/MigrateController.php [hunk 1]
```diff
@@ -196,17 +196,31 @@ protected function createMigration($class)
...
+        // sort history in desc order
+        uksort($history, function ($a, $b) {
+            if ($a == $b) {
+                return 0;
+            } else {
+                return ($a > $b) ? -1 : 1;
+            }
+        });
+        $history = ArrayHelper::map($history, 'version', 'apply_time');
         unset($history[self::BASE_MIGRATION]);
```
Comment (SilverFire) [inline framework/console/controllers/MigrateController.php]: `$history[self::BASE_MIGRATION]` is no longer a valid key. See [this](https://github.com/yiisoft/yii2/pull/12975/files#diff-952bc7045bade5b9e9c44621c5f93c19R210)

File: framework/console/controllers/MigrateController.php [hunk 2]
```diff
@@ -196,17 +196,31 @@ protected function createMigration($class)
...
         $rows = $query->select(['version', 'apply_time'])
             ->from($this->migrationTable)
-            ->orderBy('apply_time DESC, version DESC')
             ->limit($limit)
             ->createCommand($this->db)
             ->queryAll();
-        $history = ArrayHelper::map($rows, 'version', 'apply_time');
+        $history = [];
+        foreach ($rows as $row) {
+            if (preg_match('/m?(\d{6}_?\d{6})(\D.*)?$/is', $row['version'], $matches)) {
```
Comment (SilverFire) [inline framework/console/controllers/MigrateController.php]: Add here

```php
if ($row['version'] === self::BASE_MIGRATION) {
    continue;
}
```

File: framework/console/controllers/MigrateController.php [hunk 3]
```diff
@@ -196,17 +196,31 @@ protected function createMigration($class)
...
+                $time = str_replace('_', '', $matches[1]);
+                $history['m' . $time . $matches[2]] = $row;
+            }
+        }
+        // sort history in desc order
+        uksort($history, function ($a, $b) {
+            if ($a == $b) {
```
Comment (SilverFire) [inline framework/console/controllers/MigrateController.php]: `===` should be used

File: framework/console/controllers/MigrateController.php [hunk 4]
```diff
@@ -196,17 +196,31 @@ protected function createMigration($class)
...
+                return 0;
+            } else {
+                return ($a > $b) ? -1 : 1;
```
Comment (SilverFire) [inline framework/console/controllers/MigrateController.php]: use http://php.net/manual/ru/function.strcasecmp.php instead

File: framework/console/controllers/MigrateController.php [hunk 5]
```diff
@@ -196,17 +196,31 @@ protected function createMigration($class)
      */
     protected function getMigrationHistory($limit)
     {
-        if ($this->db->schema->getTableSchema($this->migrationTable, true) === null) {
+       if ($this->db->schema->getTableSchema($this->migrationTable, true) === null) {
```
Comment (SilverFire) [inline framework/console/controllers/MigrateController.php]: Gimme back that space! :hocho:

**Review by SilverFire — CHANGES_REQUESTED** [review]
Good in general. Some change should be applied before the merge though.

File: framework/console/controllers/MigrateController.php (hunk 5 above)
Comment (samdark) [inline framework/console/controllers/MigrateController.php]: ?

**Comment by evgen-d** [comment]
@SilverFire all done

**Comment by samdark** [comment]
@evgen-d tests are failing: https://travis-ci.org/yiisoft/yii2/jobs/175302999

File: framework/console/controllers/MigrateController.php [hunk 6]
```diff
@@ -202,12 +202,23 @@ protected function getMigrationHistory($limit)
...
             ->limit($limit)
             ->createCommand($this->db)
             ->queryAll();
-        $history = ArrayHelper::map($rows, 'version', 'apply_time');
-        unset($history[self::BASE_MIGRATION]);
+        $history = [];
+        foreach ($rows as $row) {
+            if ($row['version'] === self::BASE_MIGRATION) {
+                continue;
+            } elseif (preg_match('/m?(\d{6}_?\d{6})(\D.*)?$/is', $row['version'], $matches)) {
```
Comment (SilverFire) [inline framework/console/controllers/MigrateController.php]: `elseif` does not make sense. If previous condition is true, `continue` will go to next iteration of `foreach`

**Review by SilverFire — CHANGES_REQUESTED** [review]

**Comment by evgen-d** [comment]
This PR fix some problem with `migrate/down all`, but i think that `version` and `namespase` must be on different columns of DB table.
This order `orderBy('apply_time DESC, version DESC')` is wrong, and may return wrong history when you call `migrate/down [N]`

**Comment by SilverFire** [comment]
Thank you

**Comment by klimov-paul** [comment]
General note to the @yiisoft/core-developers : please use 'squash and merge' button instead of simple 'merge' whlie merging PR. This prevents GIT log to hold redundant commits.

**Pull request merged**